File: internal/api/request/printer.go
package request

import (
	"fmt"
	"io"
	"net/url"
	"sort"
	"strings"
	"text/tabwriter"
)

func PrintPackagePopularities(writer io.Writer, ppl PackagePopularityList) {
	w := tabwriter.NewWriter(writer, 15, 0, 2, ' ', 0)
	for _, pkg := range ppl.PackagePopularities {
		fmt.Fprintf(w, "%s\t%.2f\n", pkg.Name, pkg.Popularity)
	}
	w.Flush()

	if ppl.Count > 0 && ppl.Total > 0 {
		fmt.Fprintf(writer, "\n%d of %d results\n", ppl.Count, ppl.Total)
	}
}

func PrintSearchURL(writer io.Writer, baseURL string, query string) {
	if len(query) > 0 {
		fmt.Fprintf(writer, "See more results at %s/packages#query=%s\n", baseURL, query)
	}
}

func PrintShowURL(writer io.Writer, baseURL string, packages []string) {
	if len(packages) > 0 {
		for i, p := range packages {
			packages[i] = url.QueryEscape(p)
		}

		sort.StringSlice.Sort(packages)
		fmt.Fprintf(writer, "See more results at %s/compare/packages#packages=%s\n", baseURL, strings.Join(packages, ","))
	}
}

File: src/api/request/printer.rs
use super::client::PackagePopularityList;
use percent_encoding::{utf8_percent_encode, NON_ALPHANUMERIC};
use std::io::Write;
use tabwriter::TabWriter;
use url::Url;

pub fn print_package_popularities(pacakge_popularities: &PackagePopularityList) {
    let mut tw = TabWriter::new(vec![]).minwidth(15).padding(0);

    for package_popularity in &pacakge_popularities.package_popularities {
        writeln!(
            &mut tw,
            "{}\t{:.2}",
            package_popularity.name, package_popularity.popularity
        )
        .unwrap();
    }
    tw.flush().unwrap();

    print!("{}", String::from_utf8(tw.into_inner().unwrap()).unwrap());

    if pacakge_popularities.count > 0 && pacakge_popularities.total > 0 {
        println!(
            "\n{} of {} results",
            pacakge_popularities.count, pacakge_popularities.total
        );
    }
}

pub fn print_search_url(base_url: &str, query: &str) {
    if !query.is_empty() {
        println!("See more results at {}/packages#query={}", base_url, query)
    }
}

pub fn print_show_url(base_url: &str, packages: &[&str]) {
    if !packages.is_empty() {
        let mut encoded_packages: Vec<String> = packages
            .iter()
            .map(|package| utf8_percent_encode(package, NON_ALPHANUMERIC).to_string())
            .collect();
        encoded_packages.sort();

        let mut link = Url::parse(base_url).unwrap();
        link.set_path("/compare/packages");
        link.set_fragment(Some(
            format!("packages={}", encoded_packages.join(",").as_str()).as_str(),
        ));

        println!("See more results at {}", link.as_str());
    }
}
